File: go/logger.go
package main

import (
	"fmt"
	"os"
	"path/filepath"
	"strings"
	"sync"
	"syscall"
	"time"
)

const LogRetentionDays = 21

type Logger struct {
	dir       string
	file      *os.File
	mu        sync.Mutex
	currentDate string
	uid       int
	gid       int
}

func NewLogger(dir string, uid, gid int) (*Logger, error) {
	if err := os.MkdirAll(dir, 0755); err != nil {
		return nil, err
	}

	logger := &Logger{
		dir: dir,
		uid: uid,
		gid: gid,
	}

	return logger, nil
}

func (l *Logger) Write(topic string, message string) error {
	l.mu.Lock()
	defer l.mu.Unlock()

	date := getDate()
	if l.currentDate != date {
		if err := l.rotate(date); err != nil {
			return err
		}
	}

	timestamp := time.Now().Format("2006-01-02T15:04:05.000Z07:00")

	var output string
	if topic != "" {
		output = fmt.Sprintf("%s %s\n", timestamp, topic)
	} else {
		output = fmt.Sprintf("%s\n", timestamp)
	}

	// Handle multi-line messages with indentation
	lines := strings.Split(message, "\n")
	for i, line := range lines {
		if i == 0 && topic != "" {
			output += fmt.Sprintf("  %s\n", line)
		} else if line != "" {
			output += fmt.Sprintf("  %s\n", line)
		} else if i < len(lines)-1 {
			output += "\n"
		}
	}

	_, err := l.file.WriteString(output)
	if err != nil {
		return err
	}

	return l.file.Sync()
}

func (l *Logger) rotate(date string) error {
	// Close current file if open
	if l.file != nil {
		l.file.Close()
		l.file = nil
	}

	// Open new log file
	logPath := filepath.Join(l.dir, date+".log")
	file, err := os.OpenFile(logPath, os.O_CREATE|os.O_APPEND|os.O_WRONLY, 0644)
	if err != nil {
		return err
	}

	// Set ownership if running as root
	if os.Geteuid() == 0 && (l.uid > 0 || l.gid > 0) {
		syscall.Chown(logPath, l.uid, l.gid)
	}

	l.file = file
	l.currentDate = date

	// Clean up old log files
	go l.cleanup()

	return nil
}

func (l *Logger) cleanup() {
	cutoff := time.Now().AddDate(0, 0, -LogRetentionDays)

	entries, err := os.ReadDir(l.dir)
	if err != nil {
		return
	}

	for _, entry := range entries {
		if entry.IsDir() {
			continue
		}

		name := entry.Name()
		if !strings.HasSuffix(name, ".log") {
			continue
		}

		info, err := entry.Info()
		if err != nil {
			continue
		}

		if info.ModTime().Before(cutoff) {
			os.Remove(filepath.Join(l.dir, name))
		}
	}
}

func (l *Logger) Close() error {
	l.mu.Lock()
	defer l.mu.Unlock()

	if l.file != nil {
		err := l.file.Close()
		l.file = nil
		return err
	}
	return nil
}

func getDate() string {
	return time.Now().Format("2006-01-02")
}

File: src/logger.rs
use anyhow::Result;
use chrono::Local;
use std::fs::{self, File, OpenOptions};
use std::io::Write;
use std::path::PathBuf;
use std::sync::Mutex;
use std::time::SystemTime;

#[derive(Debug)]
pub struct Logger {
    dir: PathBuf,
    file: Mutex<Option<File>>,
    current_date: Mutex<String>,
    uid: u32,
    gid: u32,
    pub prune_days: i64,
}

impl Logger {
    pub fn new(dir: PathBuf, uid: u32, gid: u32, prune_days: i64) -> Result<Self> {
        fs::create_dir_all(&dir)?;

        Ok(Logger {
            dir,
            file: Mutex::new(None),
            current_date: Mutex::new(String::new()),
            uid,
            gid,
            prune_days,
        })
    }

    pub fn write(&self, topic: &str, message: &str) {
        let date = get_date();

        let mut current_date = self.current_date.lock().unwrap();
        if *current_date != date {
            let _ = self.rotate(&date);
            *current_date = date.clone();
        }
        drop(current_date);

        let timestamp = Local::now().format("%H:%M:%S");

        let msg = message.trim();
        if msg.is_empty() {
            return;
        }

        // Calculate prefix for multi-line continuation
        let prefix_len = if topic.is_empty() {
            9 // "HH:MM:SS "
        } else {
            9 + topic.len() + 3 // "HH:MM:SS [topic] "
        };
        let prefix = format!("\n{}", " ".repeat(prefix_len));

        // Format message with optional topic
        let formatted_msg = if topic.is_empty() {
            msg.replace('\n', &prefix)
        } else {
            let msg_with_topic = format!("[{}] {}", topic, msg);
            msg_with_topic.replace('\n', &prefix)
        };

        let output = format!("{} {}\n", timestamp, formatted_msg);

        let mut file = self.file.lock().unwrap();
        if let Some(ref mut f) = *file {
            let _ = f.write_all(output.as_bytes());
            let _ = f.sync_all();
        }
    }

    fn rotate(&self, date: &str) -> Result<()> {
        let mut file = self.file.lock().unwrap();

        // Close current file
        *file = None;

        // Open new log file
        let log_path = self.dir.join(format!("{}.log", date));
        let new_file = OpenOptions::new()
            .create(true)
            .append(true)
            .open(&log_path)?;

        // Set ownership if running as root
        #[cfg(target_os = "linux")]
        {
            if nix::unistd::geteuid().is_root() && (self.uid > 0 || self.gid > 0) {
                unsafe {
                    let path_cstr = std::ffi::CString::new(log_path.to_str().unwrap()).unwrap();
                    let _ = libc::chown(path_cstr.as_ptr(), self.uid, self.gid);
                }
            }
        }

        *file = Some(new_file);

        #[cfg(unix)]
        {
            // Create/update 'current' symlink
            let symlink_path = self.dir.join("current");
            let log_filename = format!("{}.log", date);
            let _ = fs::remove_file(&symlink_path); // Remove old symlink if exists
            use std::os::unix::fs::symlink;
            let _ = symlink(&log_filename, &symlink_path);
        }

        // Clean up old log files if pruning is enabled
        if self.prune_days > 0 {
            let dir = self.dir.clone();
            let prune_days = self.prune_days;
            tokio::spawn(async move {
                let _ = cleanup_old_logs(&dir, prune_days);
            });
        }

        Ok(())
    }

    #[allow(dead_code)]
    pub fn close(&self) -> Result<()> {
        let mut file = self.file.lock().unwrap();
        *file = None;
        Ok(())
    }
}

fn get_date() -> String {
    Local::now().format("%Y-%m-%d").to_string()
}

fn cleanup_old_logs(dir: &PathBuf, prune_days: i64) -> Result<()> {
    let cutoff = SystemTime::now() - std::time::Duration::from_secs((prune_days * 86400) as u64);

    let entries = fs::read_dir(dir)?;
    for entry in entries {
        let entry = entry?;
        let path = entry.path();

        if path.is_file() && path.extension().and_then(|s| s.to_str()) == Some("log") {
            if let Ok(metadata) = entry.metadata() {
                if let Ok(modified) = metadata.modified() {
                    if modified < cutoff {
                        let _ = fs::remove_file(&path);
                    }
                }
            }
        }
    }

    Ok(())
}
